### PR TITLE
MM CPU Support

### PR DESCRIPTION
**Pull request by jiaxinwu**

### CODE REVIEW

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c [hunk 1]
```diff
@@ -0,0 +1,1517 @@
...
+          MaxSize           = mSmmCpuSmramRanges[Index].PhysicalSize;
+          CurrentSmramRange = &mSmmCpuSmramRanges[Index];
+        }
+      }
+    }
+  }
+
+  ASSERT (CurrentSmramRange != NULL);
+
+  *SmrrBase = (UINT32)CurrentSmramRange->CpuStart;
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c:1352]: fixed

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuDxeSmm.c [hunk 2]
```diff
@@ -9,427 +9,116 @@
...
-    while (!(BOOLEAN)mSmmInitialized[Index]) {
-    }
-  }
+  Fadt = (EFI_ACPI_2_0_FIXED_ACPI_DESCRIPTION_TABLE *)EfiLocateFirstAcpiTable (
+                                                        EFI_ACPI_2_0_FIXED_ACPI_DESCRIPTION_TABLE_SIGNATURE
+                                                        );
+  ASSERT (Fadt != NULL);
 
-  PERF_FUNCTION_END ();
+  mSmiCommandPort = Fadt->SmiCmd;
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuDxeSmm.c:120]: not issue.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c [hunk 3]
```diff
@@ -9,6 +9,61 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+
+    //
+    // gEdkiiPiSmmMemoryAttributesTableGuid should have been published after SmmCore dispatched all MM drivers (MmDriverDispatchHandler).
+    //
+    SmmGetSystemConfigurationTable (&gEdkiiPiSmmMemoryAttributesTableGuid, (VOID **)&MemoryAttributesTable);
+
+    //
+    // Set critical region attribute in page table according to the MemoryAttributesTable
+    //
+    if (MemoryAttributesTable != NULL) {
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c:39]: When MemoryAttributesTable is NULL? Can we assert that the table is NOT NULL?

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c [hunk 4]
```diff
@@ -9,6 +9,61 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+      SetMemMapAttributes (MemoryAttributesTable);
+    }
+
+    if (IsRestrictedMemoryAccess ()) {
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c:43]: Can we remove this check?
If a platform wants to allow all DRAM access from SMM env, MmIpl should produce proper Resource HOBs to tell that all DRAM is accessible.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  gEfiMdePkgTokenSpaceGuid.PcdControlFlowEnforcementPropertyMask        ## CONSUMES
+
+[FixedPcd]
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmMpTokenCountPerChunk               ## CONSUMES
+
+[Depex]
+  TRUE
+
+[Pcd.X64]
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmRestrictedMemoryAccess        ## CONSUMES
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:139]: can we remove this?

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuHotPlugSupport                   ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmStackGuard                    ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmProfileEnable                 ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmProfileRingBuffer             ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmFeatureControlMsrLock         ## CONSUMES
+  gEfiMdeModulePkgTokenSpaceGuid.PcdDxeIplSwitchToLongMode         ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdSmmApPerfLogEnable                  ## CONSUMES
+
+[Pcd]
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuMaxLogicalProcessorNumber        ## SOMETIMES_CONSUMES
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:122]: PcdCpuMaxLogicalProcessorNumber can be a dynamic PCD. So, CpuMm should not consume it.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmProfileSize                   ## SOMETIMES_CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmStackSize                     ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmCodeAccessCheckEnable         ## CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuSmmShadowStackSize               ## SOMETIMES_CONSUMES
+  gUefiCpuPkgTokenSpaceGuid.PcdCpuFeaturesInitOnS3Resume           ## CONSUMES
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:127]: Remove it.

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c [hunk 8]
```diff
@@ -6,3 +6,42 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+    // Find gEdkiiSmmProfileDataGuid
+    //
+    if (CompareGuid (&SmmProfileDataHob.MemoryAllocation->AllocDescriptor.Name, &gEdkiiSmmProfileDataGuid)) {
+      break;
+    }
+
+    SmmProfileDataHob.Raw = GetNextHob (EFI_HOB_TYPE_MEMORY_ALLOCATION, GET_NEXT_HOB (SmmProfileDataHob));
+  }
+
+  ASSERT (SmmProfileDataHob.Raw != NULL);
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:42]: As we talked, GetSmmProfileData must be called under the condition of PcdCpuSmmProfileEnable == TRUE, in such a case, assert should be ok.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c [hunk 9]
```diff
@@ -51,6 +51,12 @@ CPU_HOT_PLUG_DATA  mCpuHotPlugData = {
   0                                             // SmrrSize
 };
 
+//
+// TRUE to indicate it's the MM_STANDALONE MM CPU driver.
+// FALSE to indicate it's the DXE_SMM_DRIVER SMM CPU driver.
+//
+BOOLEAN  mIsStandaloneMm = FALSE;
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c:58]: agree. i will do the update.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c [hunk 10]
```diff
@@ -113,3 +113,51 @@ GetMpInformationFromMpServices (
...
+{
+  EFI_STATUS  Status;
+
+  mIsStandaloneMm = TRUE;
+
+  Status =  PiSmmCpuEntryCommon ();
+
+  ASSERT_EFI_ERROR (Status);
+
+  if (FeaturePcdGet (PcdCpuSmmProfileEnable)) {
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c:197]: good idea, i agree, for stadnalonemm, we can rely on the gEdkiiSmmProfileDataGuid exist or not to decide feature enable/disable. i will talk with hongbin to align the behavior.

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmCpuMemoryManagement.c [hunk 11]
```diff
@@ -1248,37 +1242,95 @@ GenSmmPageTable (
...
+      MapAttribute.Bits.ReadWrite = 0;
+    }
+
+    if ((MemoryRegion[Index].Attribute & EFI_MEMORY_XP) != 0) {
+      if (mXdSupported) {
+        MapAttribute.Bits.Nx = 1;
+      }
+    }
+
+    GenPageTable (&PageTable, PagingMode, MemoryRegion[Index].Base, (UINTN)MemoryRegion[Index].Length, MapAttribute);
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmCpuMemoryManagement.c:1292]: agree, i will add MapMask as the parameter

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c [hunk 12]
```diff
@@ -932,6 +932,26 @@ PiSmmCpuEntryCommon (
...
+  if (RegEax <= CPUID_EXTENDED_FUNCTION) {
+    //
+    // Extended CPUID functions are not supported on this processor.
+    //
+    mXdSupported = FALSE;
+    PatchInstructionX86 (gPatchXdSupported, mXdSupported, 1);
+  }
+
+  AsmCpuid (CPUID_EXTENDED_CPU_SIG, NULL, NULL, NULL, &RegEdx);
+  if ((RegEdx & CPUID1_EDX_XD_SUPPORT) == 0) {
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c:964]: Let me email to confirm whether the feature is consistent both inside and outside SMM before we clean the code inside SMM.

if yes, we can clean the code inside smm
if not, we need update the page table after the feature is checked inside smm.

Thanks,
Jiaxin

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmCpuMemoryManagement.c [hunk 13]
```diff
@@ -1176,30 +1176,25 @@ EdkiiSmmClearMemoryAttributes (
...
 GenPageTable (
-  IN OUT UINTN        *PageTable,
-  IN     PAGING_MODE  PagingMode,
-  IN     UINT64       LinearAddress,
-  IN     UINT64       Length
+  IN OUT UINTN               *PageTable,
+  IN     PAGING_MODE         PagingMode,
+  IN     UINT64              LinearAddress,
+  IN     UINT64              Length,
+  IN     IA32_MAP_ATTRIBUTE  MapAttribute
```
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmCpuMemoryManagement.c:1186]: we have aligned offline, the name is the exitsing name of function, keep as it is.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c (hunk 3 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c:39]: as we talked offline,   MemoryAttributesTable depends on below bit check, it's not always  intsalled: EFI_MEMORY_ATTRIBUTES_RUNTIME_MEMORY_PROTECTION_NON_EXECUTABLE_PE_DATA.
i will refine the comment here.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c (hunk 4 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.c:43]: we have aligned: the function will be cleared in the furture patch to create full static pagetbale by default.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf (hunk 5 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:139]: same as above

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf (hunk 6 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:122]: agree.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf (hunk 7 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuStandaloneMm.inf:127]: agree

**Comment by jiaxinwu** [comment]
> Did you test it with SmmProfileEnable = TRUE?

not yet, existing SMM profile feature has crashed, the system can't boot when it enable. i haven't got time to check the reason. 

anyway, i will consider how to test this feature on DMR with the patch.

File: UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/PiSmmCpuCommon.c:964]: Confirmed, will remove the XD check inside SMM.

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c [hunk 14]
```diff
@@ -250,15 +250,3 @@ InitSmmS3ResumeState (
     InitSmmS3Cr3 ();
```
Comment (td36) [inline UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:250]: Above this function, the following line can be remoted.  InitSmmS3Cr3() will update SmmS3ResumeState->SmmS3Cr3
SmmS3ResumeState->SmmS3Cr3 = Cr3;

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapDxeSmm.c [hunk 15]
```diff
@@ -0,0 +1,715 @@
...
+
+  @param[in, out]     BaseAddress     Return BaseAddress of SmmProfileData.
+  @param[in, out]     Size            Return Size of SmmProfileData.
+
+  @return TRUE     SmmProfile is enabled.
+          FALSE    SmmProfile is not enabled.
+
+**/
+BOOLEAN
+IsSmmProfileEnabled (
```
Comment (td36) [inline UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapDxeSmm.c:473]: How about seperate the IsSmmProfileEnabled() into two different function for the legacy smm？Acctually it does two different things:
1.Retrive if smm profile is enabled and the smm profile base and length
2.Allocate smm profile range if enabled and Input parameters are valid

File: UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+  @param[in, out]     BaseAddress     Return BaseAddress of SmmProfileData.
+  @param[in, out]     Size            Return Size of SmmProfileData.
+
+  @return TRUE     SmmProfile is enabled.
+          FALSE    SmmProfile is not enabled.
+
+**/
+BOOLEAN
+IsSmmProfileEnabled (
```
Comment (td36) [inline UefiCpuPkg/PiSmmCpuDxeSmm/NonMmramMapStandaloneMm.c:21]: How about we only find the HOB for once and cache the result? Then we don't need to traverse the HOB list every time the function is called.

File: UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c (hunk 14 above)
Comment (jiaxinwu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/CpuS3.c:250]: Talked with Dun, we agree clean in the furture.

**Pull request closed without merging**